src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -552,44 +566,83 @@ static byte[] validateAndParseBinary(
+    if (input instanceof LocalTime) {
+      LocalTime localTime = (LocalTime) input;
+      return BigInteger.valueOf(localTime.toNanoOfDay()).divide(Power10.sb16Table[9 - scale]);
     } else if (input instanceof OffsetTime) {
-      timestamp =
-          timeStampFromLocalDateTime(
-              ((OffsetTime) input).toLocalTime().atDate(LocalDate.ofEpochDay(0)), GMT);
-    } else {
-      throw typeNotAllowedException(
-          columnName, input.getClass(), "TIME", new String[] {"String", "LocalTime", "OffsetTime"});
-    }
+      return validateAndParseTime(columnName, ((OffsetTime) input).toLocalTime(), scale);
+    } else if (input instanceof String) {
+      String stringInput = (String) input;
+      {
+        // First, try to parse LocalTime
+        LocalTime localTime = catchParsingError(() -> LocalTime.parse(stringInput));
+        if (localTime != null) {
+          return validateAndParseTime(columnName, localTime, scale);
+        }
+      }
+
+      {
+        // Alternatively, try to parse OffsetTime
+        OffsetTime offsetTime = catchParsingError((() -> OffsetTime.parse(stringInput)));
+        if (offsetTime != null) {
+          return validateAndParseTime(columnName, offsetTime.toLocalTime(), scale);
+        }
+      }
+
+      {
+        // Alternatively, try to parse integer-stored time
+        Instant parsedInstant = catchParsingError(() -> parseInstantGuessScale(stringInput));
+        if (parsedInstant != null) {
+          return validateAndParseTime(
+              columnName,
+              LocalDateTime.ofInstant(parsedInstant, ZoneOffset.UTC).toLocalTime(),
+              scale);
+        }
+      }
 
-    if (timestamp == null) {
       throw valueFormatNotAllowedException(
           columnName,
           input,
           "TIME",
           "Not a valid time, see"
-              + " https://docs.snowflake.com/en/user-guide/date-time-input-output.html#time-formats"
+              + " https://docs.snowflake.com/en/LIMITEDACCESS/snowpipe-streaming.html"
               + " for the list of supported formats");
+
+    } else {
+      throw typeNotAllowedException(
+          columnName, input.getClass(), "TIME", new String[] {"String", "LocalTime", "OffsetTime"});
     }
+  }
 
-    return timestamp
-        .getNanosSinceEpoch()
-        .toBigInteger()
-        .divide(Power10.sb16Table[9 - scale])
-        .mod(BigInteger.valueOf(24L * 60 * 60).multiply(Power10.sb16Table[scale]));
+  /**
+   * Attempts to parse integer-stored date from string input. Tries to guess the scale according to
+   * the rules documented at
+   * https://docs.snowflake.com/en/user-guide/date-time-input-output.html#auto-detection-of-integer-stored-date-time-and-timestamp-values.
+   *
+   * @param input String to parse, must represent a valid long
+   * @return Instant representing the input
+   * @throws NumberFormatException If the input in not a valid long
+   */
+  private static Instant parseInstantGuessScale(String input) {
```

Comment:
wondering if we are the first one in snowflakedb repos to guess this? there should be some commons package which handles this. WDYT?

Reply:
We are not the first to guess this scale, this logic is taken from `SnowflakeDateTimeFormat#parse`, but it is working with `SFTimestamp`, which we are trying to move away from. It would be nice to extract this logic to GSCommon, but it may take some time, so I will create a JIRA to collect all the duplicated logic we have in the SDK and move it to GSCommon.

Reply:
JIRA created: SNOW-740636